Refactor GetStreamAsync & change strategy for getting Stream of ApplicationBundle source type

petlack opened the pull request:
Hi,

I'm using MvvmCross and needed drawable resource to be placeholder. Right now, I would need to add the desired placeholder in Assets folder and set build action, which doesn't feel "native". I've extracted the code for resolving `Stream` object from path and source type into `IStreamResolver` implementations.
Implementation for `ApplicationBundle` type gets `Stream` by converting string representation of resource into integer and then gets `Stream` object by calling `Context.Resources.OpenRawResource (resourceId)`.
Example is shown in `ImageLoading.Sample` project (i.e. notice change in `Config.LoadingPlaceholderPath`).

Feel free to ask any question.

molinch commented:
Thanks for the PR @petlack :+1: 

First of all it is nice that you extracted the GetStream logic. It will be easier to maintain.
Some remarks to have this integrated:
- ApplicationBundle should continue to target Assets otherwise it will break all apps based on the lib
- You can maybe add AndroidResource as a new ImageSource? (it won't make sense on iOS but that's OK)

Could you explain the issue with ModernHttpClient too?

petlack commented:
* I agree with the ApplicationBundle
* I was thinking about adding new ImageSource type - CompiledResource - in case any other platform uses similar pattern (iOS obviously doesn't, but I don't have any experience with WP)

to the ModernHttpClient - ImageService is part of FFImageLoading.Shared project, which has no references. When I try to compile the sample project, I get following error message:
```
/Users/petlack/Dev/Xamarin/Libs/FFImageLoading/FFImageLoading.Shared/ImageService.cs(81,81): Error CS0246: The type or namespace name `ModernHttpClient' could not be found. Are you missing an assembly reference? (CS0246) (FFImageLoading.Droid)
```

petlack commented:
@molinch I've reverted the strategy for ApplicationBundle back and Added CompiledResource

Note:
* Since using strings is discouraged and using integers is not good in multiplatform shared code, I've added a cache (LRU) for caching integers.

Cheers

molinch commented:
@petlack thanks for the changes & you are right integers would be messy for cross platform

To merge that PR the iOS ImageLoaderTask should be customized to have a similar StreamResolverFactory concept. Even though there it would be more a DataResolverFactory... On iOS we return byte arrays. Using Streams doesn't seem to offer any advantage, especially when you look at NSData implementation: https://github.com/mono/maccore/blob/master/src/Foundation/NSData.cs

I personally think that the LRU might be "overkill" don't you think a simple static ConcurrentDictionary<string, int> in the CompiledResourceStreamResolver would be enough? People won't have thousands of images in the resources.

molinch commented:
@petlack I did the changes that I talked about and merged your PR
Thanks for the help!

molinch commented:
If you want to test it you can play with this Alpha Nuget
https://www.nuget.org/packages/Xamarin.FFImageLoading/1.1.6-alpha1
